xbmc/pictures/Gif.cpp

```diff
@@ -0,0 +1,617 @@
+    {
+      CLog::Log(LOGDEBUG, "Gif::PrepareTemplate(): Disposal method DISPOSE_PREVIOUS encountered, but could not find a suitable frame.");
+      return false;
+    }
+    break;
+  }
+  default:
+  {
+    CLog::Log(LOGDEBUG, "Gif::PrepareTemplate(): Unknown disposal method: %d", frame.m_disposal);
+    return false;
+  }
+  }
+  return true;
+}
+
+void Gif::SetFrameAreaToBack(unsigned char* dest, const GifFrame &frame)
+{
+  for (unsigned int dest_y = frame.m_top, src_y = 0; src_y < frame.m_height; ++dest_y, ++src_y)
+  {
+    unsigned char *to = dest + (dest_y * m_pitch) + (frame.m_left * sizeof(GifColor));
+    for (unsigned int src_x = 0; src_x < frame.m_width; ++src_x)
+    {
+      memcpy(to, &m_backColor, sizeof(m_backColor));
+      to += 4;
+    }
+  }
+}
+
+bool Gif::LoadImageFromMemory(unsigned char* buffer, unsigned int bufSize, unsigned int width, unsigned int height)
+{
+  if (!m_dll.IsLoaded())
+    return false;
+
+  if (!buffer || !bufSize || !width || !height)
+    return false;
+
+  Gifreader reader;
+  reader.buffer = buffer;
+  reader.buffSize = bufSize;
+
+  if (!Open(m_gif, (void *)&reader, ReadFromMemory))
+    return false;
+
+  if (!LoadGifMetaData(m_gif))
+    return false;
+
+  m_originalWidth = m_width;
+  m_originalHeight = m_height;
+
+  try
+  {
+    InitTemplateAndColormap();
+
+    if (!ExtractFrames(m_numFrames))
+      return false;
+  }
+  catch (std::bad_alloc& ba)
+  {
+    CLog::Log(LOGERROR, "Gif::LoadImageFromMemory(): Out of memory while extracting gif frames - %s", ba.what());
+    Release();
+    return false;
+  }
+
+  return true;
+}
+
+bool Gif::Decode(const unsigned char *pixels, unsigned int pitch, unsigned int format)
+{
+  if (m_width == 0 || m_height == 0
+    || !m_dll.IsLoaded() || !m_gif
+    || format != XB_FMT_A8R8G8B8 || !m_numFrames)
+    return false;
+
+  const unsigned char *src = m_frames[0]->m_pImage;
+  unsigned char *dst = (unsigned char*)pixels;
+
+  if (pitch == m_pitch)
+    memcpy(dst, src, m_imageSize);
+  else
+  {
+    for (unsigned int y = 0; y < m_height; y++)
+    {
+      memcpy(dst, src, m_pitch);
+      src += m_pitch;
+      dst += pitch;
+    }
+  }
+  return true;
+}
+
+bool Gif::CreateThumbnailFromSurface(unsigned char* bufferin, unsigned int width, unsigned int height, unsigned int format, unsigned int pitch, const std::string& destFile,
+                                     unsigned char* &bufferout, unsigned int &bufferoutSize)
+{
+  CLog::Log(LOGERROR, "Gif::CreateThumbnailFromSurface(): Not implemented. Something went wrong, we don't store thumbnails as gifs!");
+  return false;
+}
+
+void Gif::PrettyPrintError(std::string messageTemplate, int reason)
+{
+  const char* error = m_dll.GifErrorString(reason);
+  std::string message;
+  if (error)
+  {
+    message = StringUtils::Format(messageTemplate.append(" - %s").c_str(), error);
+  }
+  else
+  {
+    message = messageTemplate.append(" (reason unknown)");
+  }
+  CLog::Log(LOGERROR, message.c_str());
```

Comment:
ahh ok, well it's not an issue really but if it gives a warning we might as well try and change it, we don't really need more of those :)